In [37]:
from utils import setup_env_path

setup_env_path()

%load_ext autoreload
%autoreload 2

Former working directory:  /Users/ambroisebertin/Desktop/prog/prog_abeilles/fil-rouge-pollinisateurs
Current working directory:  /Users/ambroisebertin/Desktop/prog/prog_abeilles/fil-rouge-pollinisateurs
The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [77]:
import pandas as pd
import numpy as np
from tqdm import tqdm
import missingno as msno

In [39]:
API_keys = [
    "2b10Me1HF0rfjoGWCseolNa8e",
    "2b10Xs7brWPuBdRTSeWx9V7HJu",
    "2b10umm8L2jTYjlWrJKPfshJ0O",
    "2b10sEg5pSsrtVT372XqlrzcLe",
    "2b10LDWWlXJwuS6FleaqfK6Rke",
]

## Préparation des données

In [40]:
# Import spipoll.csv as a pandas dataframe
spipoll = pd.read_csv("././data/spipoll.csv",low_memory=False)

# Extract relevant columns from the spipoll dataframe
plantes = spipoll[['collection_id', 'plante_sc', 'plante_fr',
       'plante_precision', 'plante_inconnue', 'plante_caractere',
       'photo_fleur', 'photo_plante', 'photo_feuille']]

# Shrinking the data by grouping by collection_id
# Keep only the first row for each unique value of collection_id
plantes = plantes.drop_duplicates(subset='collection_id', keep='first')

plantes.shape

(75372, 9)

In [41]:
values_to_check = ["Je ne sais pas", "Plante inconnue"]

In [42]:
plantes_sc_unlabelled = plantes.loc[ #renommer en plantes_unlabelled_sc pour plus de clarté ??
    (
        (plantes["plante_sc"].isna() | plantes["plante_sc"].isin(values_to_check))
    )
    | ((plantes["plante_inconnue"] == 1.0))
]

# creation du dataframe contenant les valeurs de plantes mais pas de plantes_sc_unlabelled
plantes_sc_labelled = plantes[~plantes.index.isin(plantes_sc_unlabelled.index)]

# in plante add "1" to the column group for each row that is in plantes_sc_labelled
plantes.loc[plantes_sc_labelled.index, 'group'] = 1

plantes_sc_labelled.shape

(63465, 9)

In [43]:
# verification des dimensions des deux df par rapport à la dim de plantes
plantes_sc_labelled.shape[0] + plantes_sc_unlabelled.shape[0] == plantes.shape[0]

True

Compression de l'information étiquettée redondante :

In [44]:
# group plantes_sc_labelled by unique values of plante_sc
plantes_sc_labelled = plantes_sc_labelled.drop_duplicates(subset='plante_sc', keep='first')

# display the dimension of the dataframe
plantes_sc_labelled.shape

(2620, 9)

On a donc un df avec 2620 plantes dont le nom est spécifié dans plante_sc et 11.907 (~1.8% des données d'origine, spipoll.csv) dont le nom n'est pas spécifié dans plante_sc.

In [45]:
# ajoute trois colonnes après plante_sc dans le dataframe plantes_sc_labelled "Famille", "Genre" et "Espèce" entre la colonne plante_sc et plante_fr
plantes_sc_labelled.insert(2, "Famille", np.nan)
plantes_sc_labelled.insert(3, "Genre", np.nan)
plantes_sc_labelled.insert(4, "Espece", np.nan)

# ajoute trois colonnes après plante_sc dans le dataframe plantes_sc_unlabelled "Famille", "Genre" et "Espèce" entre la colonne plante_sc et plante_fr
plantes_sc_unlabelled.insert(2, "Famille", np.nan)
plantes_sc_unlabelled.insert(3, "Genre", np.nan)
plantes_sc_unlabelled.insert(4, "Espece", np.nan)

In [46]:
plantes_sc_labelled.head()

collection_id                                          plante_sc  Famille  \
0              1  Les Trèfles à fleurs blanches ou roses en boul...      NaN   
2              2                                    Acanthus mollis      NaN   
5              3  Les Orchidées à fleurs blanches, jaunes ou ver...      NaN   
6              4  Les Vipérines à fleurs bleues-violacées (des e...      NaN   
10             5                        Les Pâquerettes (Bellis sp)      NaN   

    Genre  Espece      plante_fr                       plante_precision  \
0     NaN     NaN            NaN                                    NaN   
2     NaN     NaN  Acanthe molle                                    NaN   
5     NaN     NaN            NaN  orchis bouc - Himantoglossum hircinum   
6     NaN     NaN            NaN                               Theaceae   
10    NaN     NaN            NaN                                    NaN   

    plante_inconnue plante_caractere  \
0               0.0        spontanée   
2               NaN          plantée   
5               0.0        spontanée   
6               0.0          plantée   
10              0.0        spontanée   

                                          photo_fleur  \
0   https://spgp-api.65mo.fr/api/containers/spgp/d...   
2   https://spgp-api.65mo.fr/api/containers/spgp/d...   
5   https://spgp-api.65mo.fr/api/containers/spgp/d...   
6   https://spgp-api.65mo.fr/api/containers/spgp/d...   
10  https://spgp-api.65mo.fr/api/containers/spgp/d...   

                                         photo_plante  \
0   https://spgp-api.65mo.fr/api/containers/spgp/d...   
2   https://spgp-api.65mo.fr/api/containers/spgp/d...   
5   https://spgp-api.65mo.fr/api/containers/spgp/d...   
6   https://spgp-api.65mo.fr/api/containers/spgp/d...   
10  https://spgp-api.65mo.fr/api/containers/spgp/d...   

                                        photo_feuille  
0   https://spgp-api.65mo.fr/api/containers/spgp/d...  
2   https://spgp-api.65mo.fr/api/containers/spgp/d...  
5   https://spgp-api.65mo.fr/api/containers/spgp/d...  
6   https://spgp-api.65mo.fr/api/containers/spgp/d...  
10  https://spgp-api.65mo.fr/api/containers/spgp/d...

## Appel à l'API

In [47]:
from data_quality.plant_treatment.plantnet_api import PlantNetPredictor

def call_API(iloc, api_key_index=0):
    predictor = PlantNetPredictor(key = API_keys[api_key_index])
    prediction = None

    try:
        prediction = predictor.predict(imageURL= plantes_sc_labelled['photo_fleur'].iloc[iloc],
                        organs = "auto",
                        includeRelatedImages=False)
    except Exception as e:
        print(f"Error: {e}")
        if api_key_index + 1 < len(API_keys):
            return get_plant_details(iloc, api_key_index + 1)
        else:
            print("You used all your API keys for today.")
            return None

    famille = prediction["results"][0]["species"]["family"]["scientificNameWithoutAuthor"] # famille
    genre = prediction["results"][0]["species"]["genus"]["scientificNameWithoutAuthor"] # genre
    espece = prediction["results"][0]["species"]["scientificNameWithoutAuthor"] # espece

    return [famille, genre, espece]

In [84]:
def remplir_tableau(df, num):
    start_index = 0
    for i, val in enumerate(df["Famille"]):
        if pd.isnull(val):
            start_index = i
            break
    else:
        print("Le dataframe est déjà rempli.")
        return df

    for i in tqdm(range(start_index, start_index + num)):
        try:
            df.iloc[i, 2:5] = call_API(i)
        except Exception as e:
            print(f"Error: {e}")

    return df

In [85]:
remplir_tableau(plantes_sc_labelled, 7)

  0%|          | 0/7 [00:00<?, ?it/s]

100%|██████████| 7/7 [00:21<00:00,  3.14s/it]


collection_id                                          plante_sc  \
0                  1  Les Trèfles à fleurs blanches ou roses en boul...   
2                  2                                    Acanthus mollis   
5                  3  Les Orchidées à fleurs blanches, jaunes ou ver...   
6                  4  Les Vipérines à fleurs bleues-violacées (des e...   
10                 5                        Les Pâquerettes (Bellis sp)   
...              ...                                                ...   
665235        75 468                                  Arbutus andrachne   
665583        75 525                                    Pavonia hastata   
665861        75 554                                   Medicago falcata   
666544        75 625                             Helianthus divaricatus   
666854        75 657                                  Sideritis hirsuta   

            Famille           Genre                   Espece  \
0          Fabaceae       Trifolium       Trifolium pratense   
2       Acanthaceae        Acanthus        Acanthus spinosus   
5       Orchidaceae  Himantoglossum  Himantoglossum hircinum   
6          Theaceae        Camellia        Camellia sasanqua   
10       Asteraceae          Bellis        Bellis sylvestris   
...             ...             ...                      ...   
665235          NaN             NaN                      NaN   
665583          NaN             NaN                      NaN   
665861          NaN             NaN                      NaN   
666544          NaN             NaN                      NaN   
666854          NaN             NaN                      NaN   

                           plante_fr                       plante_precision  \
0                                NaN                                    NaN   
2                      Acanthe molle                                    NaN   
5                                NaN  orchis bouc - Himantoglossum hircinum   
6                                NaN                               Theaceae   
10                               NaN                                    NaN   
...                              ...                                    ...   
665235           Arbousier de Chypre                                    NaN   
665583     Pavonie (Pavonia hastata)                                    NaN   
665861           Luzerne en faucille                                    NaN   
666544  Hélianthe à feuilles Ètalées                                    NaN   
666854           Crapaudine d'Endres                                    NaN   

        plante_inconnue plante_caractere  \
0                   0.0        spontanée   
2                   NaN          plantée   
5                   0.0        spontanée   
6                   0.0          plantée   
10                  0.0        spontanée   
...                 ...              ...   
665235              0.0        spontanée   
665583              NaN          plantée   
665861              0.0        spontanée   
666544              NaN        spontanée   
666854              NaN          plantée   

                                              photo_fleur  \
0       https://spgp-api.65mo.fr/api/containers/spgp/d...   
2       https://spgp-api.65mo.fr/api/containers/spgp/d...   
5       https://spgp-api.65mo.fr/api/containers/spgp/d...   
6       https://spgp-api.65mo.fr/api/containers/spgp/d...   
10      https://spgp-api.65mo.fr/api/containers/spgp/d...   
...                                                   ...   
665235  https://spgp-api.65mo.fr/api/containers/spgp/d...   
665583  https://spgp-api.65mo.fr/api/containers/spgp/d...   
665861  https://spgp-api.65mo.fr/api/containers/spgp/d...   
666544  https://spgp-api.65mo.fr/api/containers/spgp/d...   
666854  https://spgp-api.65mo.fr/api/containers/spgp/d...   

                                             photo_plante  \
0       https://spgp-api.65mo.fr/api/containers/spgp/d...   
2   

In [86]:
plantes_sc_labelled.head(16) # verification des deux fois 7 premières lignes

collection_id                                          plante_sc  \
0               1  Les Trèfles à fleurs blanches ou roses en boul...   
2               2                                    Acanthus mollis   
5               3  Les Orchidées à fleurs blanches, jaunes ou ver...   
6               4  Les Vipérines à fleurs bleues-violacées (des e...   
10              5                        Les Pâquerettes (Bellis sp)   
24              6  Les Bruyères à fleurs roses en grappes (des es...   
25              7             Les Ajoncs (des espèces du genre Ulex)   
41             12                      Les Pissenlits (Taraxacum sp)   
47             15  Les Crucifères jaunes (des espèces de la famil...   
54             17                                   Senecio jacobaea   
55             18  Les Viornes (à fleurs semblables) (des espèces...   
56             19                             Rosmarinus officinalis   
62             20  Les Abricotiers, Pruniers et Amandiers (à fleu...   
81             27  Les Bruyères à fleurs blanches (des espèces du...   
90             29  Les Sénéçons sans fleurs ligulées (des espèces...   
108            36  Les Oxalis à fleurs jaunes (des espèces du gen...   

          Famille           Genre                   Espece  \
0        Fabaceae       Trifolium       Trifolium pratense   
2     Acanthaceae        Acanthus        Acanthus spinosus   
5     Orchidaceae  Himantoglossum  Himantoglossum hircinum   
6        Theaceae        Camellia        Camellia sasanqua   
10     Asteraceae          Bellis        Bellis sylvestris   
24      Ericaceae           Erica      Erica x darleyensis   
25       Fabaceae     Adenocarpus   Adenocarpus foliolosus   
41     Asteraceae       Taraxacum       Taraxacum palustre   
47   Brassicaceae         Sinapis         Sinapis arvensis   
54     Asteraceae         Senecio      Senecio inaequidens   
55      Adoxaceae        Viburnum           Viburnum tinus   
56      Lamiaceae      Rosmarinus   Rosmarinus officinalis   
62       Rosaceae          Prunus           Prunus spinosa   
81      Ericaceae           Erica      Erica x darleyensis   
90            NaN             NaN                      NaN   
108           NaN             NaN                      NaN   

              plante_fr                       plante_precision  \
0                   NaN                                    NaN   
2         Acanthe molle                                    NaN   
5                   NaN  orchis bouc - Himantoglossum hircinum   
6                   NaN                               Theaceae   
10                  NaN                                    NaN   
24                  NaN          Bruyère carnée (Erica carnea)   
25                  NaN                 Ajonc (Ulex europaeus)   
41                  NaN       Pissenlit (Taraxacum officinale)   
47                  NaN                                    NaN   
54   Le Sénéçon jacobée                                    NaN   
55                  NaN                                    NaN   
56           Le Romarin                                    NaN   
62                  NaN                     Prunellier sauvage   
81                  NaN          Bruyère carnée (Erica carnea)   
90                  NaN                                    NaN   
108                 NaN                                    NaN   

     plante_inconnue    plante_caractere  \
0                0.0           spontanée   
2                NaN             plantée   
5                0.0           spontanée   
6                0.0             plantée   
10               0.0           spontanée   
24               0.0             plantée   
25               0.0  ne se prononce pas   
41               0.0           spontanée   
47               0.0           spontanée   
54               0.0           spontanée   
55               0.0             plantée   
56               0.0             plantée   
62               0.0         

In [66]:
# (uniquement utile pour faire des tests) vide les colonnes "Famille", "Genre" et "Espece" du dataframe plantes_sc_labelled
plantes_sc_labelled.iloc[:, 2:5] = np.nan

In [89]:
plantes_sc_labelled.shape # verification du maintien de la dimension du dataframe : OK

(2620, 12)

## Pipeline